### PR TITLE
Refactor layer flushing task

### PR DESCRIPTION
**Pull request by LizardWizzard**
Not the biggest part of on-demand PR #2595  but at least a separate commit that can be reviewed separately. Did a cherry-pick on top of main + resolved conflicts.

Also reducing search space for bugs in ondemand PR

WIP: running tests

### CODE REVIEW

File: pageserver/src/tenant.rs [hunk 1]
```diff
@@ -978,81 +985,88 @@ impl Tenant {
...
+            let timeline_ids = {
+                if let Some(target_timeline_id) = target_timeline_id.as_ref() {
+                    if timelines.get(target_timeline_id).is_none() {
+                        bail!("gc target timeline does not exist")
+                    }
+                };
+
+                timelines
+                    .iter()
+                    // TODO here was the filter to take only active timelines. Is it actually correct?
```
Comment (LizardWizzard) [inline pageserver/src/tenant.rs:1006]: It is out of scope for the PR but just a change I've noticed. Here in main we filter inactive timelines which I think is wrong. Gc needs to know about all branch points, not only ones for timelines that are active at the moment of gc. If timeline is inactive then we wont know about branch point and I believe in this case gc can delete data that is needed by child timeline.

Comment (hlinnaka) [reply pageserver/src/tenant.rs:1030]: Agreed

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -788,6 +743,22 @@ impl Timeline {
         result
     }
 
+    pub(super) fn spawn_flush_loop(self: &Arc<Self>) {
```
Comment (hlinnaka) [inline pageserver/src/tenant/timeline.rs:746]: Let's add a boolean to Timeline struct, to keep track of whether the flush loop has already been started. If we start two flush loops on the same timeline, bad things will happen, and it's currently not very clear when it's started or who's responsible for starting it. Would be good to make that more clear, but a boolean would go a long way. We could then just not start it again if its already running, which would allow you to just always call spawn_flush_loop() if it might or might not be running. In particular, `initialize_with_lock` could just always call spawn_flush_loop(), it wouldn't need the new `spawn_flush_loop` argument.

**Review by hlinnaka — COMMENTED** [review]
thanks for extracting this!

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (LizardWizzard) [reply pageserver/src/tenant/timeline.rs:746]: Yeah, flush_loop supervision is one of my concerns. It is possible for it to stop due to panic (not that high of a probability, but still) and in this case checkpointer wont be able to make progress. I thought about adding bool to timeline to avoid double start problem but decided against it since it was only two cases and in other ones flag is equal to load_layer_map one we already have there. Additionally `initialize` without flags is also used for "default" cases. So I moved manual invocations to flush_loop and call to checkpoint to the same function so it is easier to see such places. I agree that having a bool would be safer. Will change. I also think that double start should be considered as an error (debug_assert maybe?)

Ideally we would rather store a join handle to check if task is alive and restart it when it isnt. Or add this logic to the task manager (aka restart_on_error flag to task_mgr::spawn)

Comment (LizardWizzard) [reply pageserver/src/tenant/timeline.rs:746]: Hmm, remembered another reason why I did it this way. If we start flush_loop only in `initialize()` then checkpoint needs to happen after that, so we register timeline as ready and observable by `get_timeline` calls before we actually finish the whole import routine (is it correct to treat checkpoint as a part of this process?). We also start listening for wal_receiver events.

Is it a real concern or its ok to do it either way?

**Comment by LizardWizzard** [comment]
Had weird failure in tests only for release build. I wasnt able to reproduce it locally. Not sure if it has something to do with the PR

```
2022-10-28T17:18:58.886374Z  INFO handle_pagerequests{tenant_id=298aa909f0574f3191d3b171018011e0 timeline_id=4c48715d22949ffe476ae9c46b7c00ca}:handle_get_page_at_lsn_request{rel=1663/12972/2663 blkno=1 req_lsn=0/16A6660}: old temporary datadir /tmp/test_output/test_pageserver_recovery/repo/tenants/298aa909f0574f3191d3b171018011e0/wal-redo-datadir.___temp exists, removing
2022-10-28T17:18:58.996326Z ERROR handle_pagerequests{tenant_id=298aa909f0574f3191d3b171018011e0 timeline_id=4c48715d22949ffe476ae9c46b7c00ca}: error reading relation or page version: Directory not empty (os error 39)

Stack backtrace:
   0: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/result.rs:2106:27
      pageserver::tenant::timeline::Timeline::reconstruct_value
             at /__w/neon/neon/pageserver/src/tenant/timeline.rs:2292:27
   1: pageserver::tenant::timeline::Timeline::get::{{closure}}
             at /__w/neon/neon/pageserver/src/tenant/timeline.rs:374:42
      prometheus::histogram::Histogram::observe_closure_duration
             at /github/home/.cargo/registry/src/github.com-1ecc6299db9ec823/prometheus-0.13.2/src/histogram.rs:716:19
```

File: pageserver/src/tenant.rs (hunk 1 above)
Comment (LizardWizzard) [reply pageserver/src/tenant.rs]: Fixed in separate PR #2726

**Review by SomeoneToIgnore — APPROVED** [review]
I'm a bit worried by the increased `Arc<Timeline>` amount, but guess we'll have to wait for https://github.com/neondatabase/neon/issues/2705 and its continuation on the timelines.

**Comment by LizardWizzard** [comment]
> I'm a bit worried by the increased Arc<Timeline> amount

In this case there is no other way around because we want to spawn flush loop and give it its own reference to tenant

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (LizardWizzard) [reply pageserver/src/tenant/timeline.rs]: Settled with flag option. @hlinnaka ping, if it looks ok to you lets merge it :)

File: pageserver/src/walredo.rs [hunk 3]
```diff
@@ -120,7 +120,7 @@ fn can_apply_in_neon(rec: &NeonWalRecord) -> bool {
 /// An error happened in WAL redo
 #[derive(Debug, thiserror::Error)]
 pub enum WalRedoError {
-    #[error(transparent)]
+    #[error("encountered io error: {0}")]
```
Comment (hlinnaka) [inline pageserver/src/walredo.rs:123]: This seems unrelated

**Review by hlinnaka — APPROVED** [review]
Looks good, thanks!

File: pageserver/src/tenant.rs [hunk 4]
```diff
@@ -200,37 +207,44 @@ impl UninitializedTimeline<'_> {
...
-                    "Failed to import basebackup for timeline {}/{}",
-                    self.owning_tenant.tenant_id, self.timeline_id
-                )
-            },
-        )?;
+
+        // import_basebackup_from_tar() is not async, mainly because the Tar crate
+        // it uses is not async. So we need to jump through some hoops:
+        // - convert the input from client connection to a synchronous Read
+        // - use block_in_place()
```
Comment (problame) [inline pageserver/src/tenant.rs:220]: I think our current policy is to use block_in_place if it's local disk IO, since we assume that it's fast.
But our block_in_place below will read from the SyncIoBridge'd client connection, i.e., network IO.

Hence, I think `spawn_blocking` would be more appropriate.

File: pageserver/src/tenant.rs [hunk 5]
```diff
@@ -200,37 +207,44 @@ impl UninitializedTimeline<'_> {
...
+        let reader = SyncIoBridge::new(StreamReader::new(&mut copyin_stream));
+
+        tokio::task::block_in_place(|| {
+            import_datadir::import_basebackup_from_tar(raw_timeline, reader, base_lsn)
+                .context("Failed to import basebackup")
```
Comment (problame) [inline pageserver/src/tenant.rs:225]: We lost the details in the error message `(tenant_id, timeline_id)`

Same thing below in the `.checkpoint()`

File: pageserver/src/tenant/timeline.rs [hunk 6]
```diff
@@ -1267,53 +1254,128 @@ impl Timeline {
...
+    async fn flush_frozen_layers_and_wait(&self) -> anyhow::Result<()> {
+        let mut rx = self.layer_flush_done_tx.subscribe();
 
-        Ok(())
+        // Increment the flush cycle counter and wake up the flush task.
+        // Remember the new value, so that when we listen for the flush
+        // to finish, we know when the flush that we initiated has
+        // finished, instead of some other flush that was started earlier.
+        let mut my_flush_request = 0;
+        self.layer_flush_start_tx.send_modify(|counter| {
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:1347]: We should assert that the flush loop is running. 
Perhaps just check `self.flush_loop_started`.

File: pageserver/src/tenant/timeline.rs [hunk 7]
```diff
@@ -1267,53 +1254,128 @@ impl Timeline {
...
+            my_flush_request = *counter + 1;
+            *counter = my_flush_request;
+        });
+
+        loop {
+            {
+                let (last_result_counter, last_result) = &*rx.borrow();
+                if *last_result_counter >= my_flush_request {
+                    if let Err(_err) = last_result {
+                        // We already logged the original error in
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:1357]: Can't we at least log the counter value so that we can correlate the two log messages?

**Review by problame — CHANGES_REQUESTED** [review]
Mostly small nitpicks.
Also, here's a patch on top of this to reduce the diff.
(I didn't want to push into your branch.)

[0001-reduce-diff-move-check_checkpoint_distance-back-to-o.patch.txt](https://github.com/neondatabase/neon/files/9926781/0001-reduce-diff-move-check_checkpoint_distance-back-to-o.patch.txt)

File: pageserver/src/walredo.rs (hunk 3 above)
Comment (LizardWizzard) [reply pageserver/src/walredo.rs:123]: That was helpful during local debugging, I can extract it to separate PR. With transparent it will just yield the io Error without mentioning WalRedoError at all

File: pageserver/src/tenant.rs (hunk 5 above)
Comment (LizardWizzard) [reply pageserver/src/tenant.rs:225]: It is included in the span fields. No need to duplicate this in the message

File: pageserver/src/tenant.rs (hunk 4 above)
Comment (LizardWizzard) [reply pageserver/src/tenant.rs:220]: This is not introduced by this PR. Here I just moved this code from one function to another. See the discussion in the original [PR](https://github.com/neondatabase/neon/pull/2376/files#r962137730)

`block_in_place` is similar in many ways to `spawn_blocking`. And it is different from just blocking current thread as we do with disk operations. We indeed block in place when doing disk operations, but we're not using `block_in_place` :)

Still I agree that `spawn_blocking` fits better. Block in place is still ok but rather for background runtime, not the one we use for request handling. So any objections to replacing it separetely?

**Comment by LizardWizzard** [comment]
> Also, here's a patch on top of this to reduce the diff.
(I didn't want to push into your branch.)

Thanks! It is a rebase artifact, there was a patch that moved functions around and I missed it.

File: pageserver/src/tenant/timeline.rs (hunk 7 above)
Comment (LizardWizzard) [reply pageserver/src/tenant/timeline.rs:1360]: Done

File: pageserver/src/tenant/timeline.rs (hunk 6 above)
Comment (LizardWizzard) [reply pageserver/src/tenant/timeline.rs:1350]: Done

File: pageserver/src/tenant.rs (hunk 4 above)
Comment (problame) [reply pageserver/src/tenant.rs:220]: No, go ahead

Comment (LizardWizzard) [reply pageserver/src/tenant.rs:225]: At some point I think we'll move to async disk access. 

I'll try to fit spawn_blocking here after merge of this PR

**Comment by LizardWizzard** [comment]
Got weird CI error. Looks like this issue: https://github.com/neondatabase/neon/issues/467

**Review by problame — APPROVED** [review]
